File: java/pdfclown.lib/src/org/pdfclown/documents/interaction/annotations/Border.java
/*
  Copyright 2008-2015 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library"
  (the Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

package org.pdfclown.documents.interaction.annotations;

import org.pdfclown.PDF;
import org.pdfclown.VersionEnum;
import org.pdfclown.documents.Document;
import org.pdfclown.documents.contents.LineDash;
import org.pdfclown.objects.PdfArray;
import org.pdfclown.objects.PdfDictionary;
import org.pdfclown.objects.PdfDirectObject;
import org.pdfclown.objects.PdfName;
import org.pdfclown.objects.PdfNumber;
import org.pdfclown.objects.PdfObjectWrapper;
import org.pdfclown.objects.PdfReal;

/**
  Border characteristics [PDF:1.6:8.4.3].

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.7
  @version 0.1.2.1, 03/21/15
*/
@PDF(VersionEnum.PDF11)
public final class Border
  extends PdfObjectWrapper<PdfDictionary>
{
  // <class>
  // <classes>
  /**
    Border style [PDF:1.6:8.4.3].
  */
  public enum StyleEnum
  {
    /**
      Solid.
    */
    Solid(PdfName.S),
    /**
      Dashed.

      @see #getPattern()
    */
    Dashed(PdfName.D),
    /**
      Beveled.
    */
    Beveled(PdfName.B),
    /**
      Inset.
    */
    Inset(PdfName.I),
    /**
      Underline.
    */
    Underline(PdfName.U);

    /**
      Gets the style corresponding to the given value.
    */
    public static StyleEnum get(
      PdfName value
      )
    {
      for(StyleEnum style : StyleEnum.values())
      {
        if(style.getCode().equals(value))
          return style;
      }
      return null;
    }

    private final PdfName code;

    private StyleEnum(
      PdfName code
      )
    {this.code = code;}

    public PdfName getCode(
      )
    {return code;}
  }
  // </classes>

  // <static>
  // <fields>
  private static final LineDash DefaultLineDash = new LineDash(new double[]{3});
  private static final StyleEnum DefaultStyle = StyleEnum.Solid;
  private static final double DefaultWidth = 1;
  // </fields>
  // </static>

  // <dynamic>
  // <constructors>
  /**
    Creates a non-reusable instance.
  */
  public Border(
    double width
    )
  {this(null, width);}
  
  /**
    Creates a non-reusable instance.
  */
  public Border(
    double width,
    StyleEnum style
    )
  {this(null, width, style);}
  
  /**
    Creates a non-reusable instance.
  */
  public Border(
    double width,
    LineDash pattern
    )
  {this(null, width, pattern);}
  
  /**
    Creates a reusable instance.
  */
  public Border(
    Document context,
    double width
    )
  {this(context, width, null, null);}
  
  /**
    Creates a reusable instance.
  */
  public Border(
    Document context,
    double width,
    StyleEnum style
    )
  {this(context, width, style, null);}
  
  /**
    Creates a reusable instance.
  */
  public Border(
    Document context,
    double width,
    LineDash pattern
    )
  {this(context, width, StyleEnum.Dashed, pattern);}
  
  private Border(
    Document context,
    double width,
    StyleEnum style,
    LineDash pattern
    )
  {
    super(
      context,
      new PdfDictionary(
        new PdfName[]
        {PdfName.Type},
        new PdfDirectObject[]
        {PdfName.Border}
        )
      );
    setWidth(width);
    setStyle(style);
    setPattern(pattern);
  }

  Border(
    PdfDirectObject baseObject
    )
  {super(baseObject);}
  // </constructors>

  // <interface>
  // <public>
  @Override
  public Border clone(
    Document context
    )
  {return (Border)super.clone(context);}

  /**
    Gets the dash pattern used in case of dashed border.
  */
  public LineDash getPattern(
    )
  {
    PdfArray dashObject = (PdfArray)getBaseDataObject().get(PdfName.D);
    return dashObject != null ? LineDash.get(dashObject, null) : DefaultLineDash;
  }

  /**
    Gets the border style.
  */
  public StyleEnum getStyle(
    )
  {
    PdfName styleObject = (PdfName)getBaseDataObject().get(PdfName.S);
    return styleObject != null ? StyleEnum.get(styleObject) : DefaultStyle;
  }

  /**
    Gets the border width in points.
  */
  public double getWidth(
    )
  {
    PdfNumber<?> widthObject = (PdfNumber<?>)getBaseDataObject().get(PdfName.W);
    return widthObject != null ? widthObject.getDoubleValue() : DefaultWidth;
  }

  /**
    @see #getPattern()
  */
  public void setPattern(
    LineDash value
    )
  {
    PdfArray dashObject = null;
    if(value != null)
    {
      dashObject = new PdfArray();
      for(double dashItem : value.getDashArray())
      {dashObject.add(PdfReal.get(dashItem));}
    }
    getBaseDataObject().put(PdfName.D, dashObject);
  }

  /**
    @see #getStyle()
  */
  public void setStyle(
    StyleEnum value
    )
  {getBaseDataObject().put(PdfName.S, value != null && value != DefaultStyle ? value.getCode() : null);}

  /**
    @see #getWidth()
  */
  public void setWidth(
    double value
    )
  {getBaseDataObject().put(PdfName.W, PdfReal.get(value));}
  // </public>
  // </interface>
  // </dynamic>
  // </class>
}

File: PdfClown-NetStandard/PdfClown/org/pdfclown/documents/interaction/annotations/Border.cs
/*
  Copyright 2008-2015 Stefano Chizzolini. http://www.pdfclown.org

  Contributors:
    * Stefano Chizzolini (original code developer, http://www.stefanochizzolini.it)

  This file should be part of the source code distribution of "PDF Clown library" (the
  Program): see the accompanying README files for more info.

  This Program is free software; you can redistribute it and/or modify it under the terms
  of the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3 of the License, or (at your option) any later version.

  This Program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY,
  either expressed or implied; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.

  You should have received a copy of the GNU Lesser General Public License along with this
  Program (see README files); if not, go to the GNU website (http://www.gnu.org/licenses/).

  Redistribution and use, with or without modification, are permitted provided that such
  redistributions retain the above copyright notice, license and disclaimer, along with
  this list of conditions.
*/

using org.pdfclown.bytes;
using org.pdfclown.documents;
using org.pdfclown.documents.contents;
using org.pdfclown.objects;

using System;
using System.Collections.Generic;

namespace org.pdfclown.documents.interaction.annotations
{
  /**
    <summary>Border characteristics [PDF:1.6:8.4.3].</summary>
  */
  [PDF(VersionEnum.PDF11)]
  public sealed class Border
    : PdfObjectWrapper<PdfDictionary>
  {
    #region types
    /**
      <summary>Border style [PDF:1.6:8.4.3].</summary>
    */
    public enum StyleEnum
    {
      /**
        <summary>Solid.</summary>
      */
      Solid,
      /**
        <summary>Dashed.</summary>
      */
      Dashed,
      /**
        <summary>Beveled.</summary>
      */
      Beveled,
      /**
        <summary>Inset.</summary>
      */
      Inset,
      /**
        <summary>Underline.</summary>
      */
      Underline
    };
    #endregion

    #region static
    #region fields
    private static readonly LineDash DefaultLineDash = new LineDash(new double[]{3});
    private static readonly StyleEnum DefaultStyle = StyleEnum.Solid;
    private static readonly double DefaultWidth = 1;

    private static readonly Dictionary<StyleEnum,PdfName> StyleEnumCodes;
    #endregion

    #region constructors
    static Border()
    {
      StyleEnumCodes = new Dictionary<StyleEnum,PdfName>();
      StyleEnumCodes[StyleEnum.Solid] = PdfName.S;
      StyleEnumCodes[StyleEnum.Dashed] = PdfName.D;
      StyleEnumCodes[StyleEnum.Beveled] = PdfName.B;
      StyleEnumCodes[StyleEnum.Inset] = PdfName.I;
      StyleEnumCodes[StyleEnum.Underline] = PdfName.U;
    }
    #endregion

    #region interface
    #region private
    /**
      <summary>Gets the code corresponding to the given value.</summary>
    */
    private static PdfName ToCode(
      StyleEnum value
      )
    {return StyleEnumCodes[value];}

    /**
      <summary>Gets the style corresponding to the given value.</summary>
    */
    private static StyleEnum ToStyleEnum(
      PdfName value
      )
    {
      foreach(KeyValuePair<StyleEnum,PdfName> style in StyleEnumCodes)
      {
        if(style.Value.Equals(value))
          return style.Key;
      }
      return DefaultStyle;
    }
    #endregion
    #endregion
    #endregion

    #region dynamic
    #region constructors
    /**
      <summary>Creates a non-reusable instance.</summary>
    */
    public Border(
      double width
      ) : this(null, width)
    {}

    /**
      <summary>Creates a non-reusable instance.</summary>
    */
    public Border(
      double width,
      StyleEnum style
      ) : this(null, width, style)
    {}

    /**
      <summary>Creates a non-reusable instance.</summary>
    */
    public Border(
      double width,
      LineDash pattern
      ) : this(null, width, pattern)
    {}

    /**
      <summary>Creates a reusable instance.</summary>
    */
    public Border(
      Document context,
      double width
      ) : this(context, width, DefaultStyle, null)
    {}

    /**
      <summary>Creates a reusable instance.</summary>
    */
    public Border(
      Document context,
      double width,
      StyleEnum style
      ) : this(context, width, style, null)
    {}

    /**
      <summary>Creates a reusable instance.</summary>
    */
    public Border(
      Document context,
      double width,
      LineDash pattern
      ) : this(context, width, StyleEnum.Dashed, pattern)
    {}

    private Border(
      Document context,
      double width,
      StyleEnum style,
      LineDash pattern
      ) : base(
        context,
        new PdfDictionary(
          new PdfName[]
          {PdfName.Type},
          new PdfDirectObject[]
          {PdfName.Border}
          )
        )
    {
      Width = width;
      Style = style;
      Pattern = pattern;
    }

    internal Border(
      PdfDirectObject baseObject
      ) : base(baseObject)
    {}
    #endregion

    #region interface
    #region public
    /**
      <summary>Gets/Sets the dash pattern used in case of dashed border.</summary>
    */
    public LineDash Pattern
    {
      get
      {
        PdfArray dashObject = (PdfArray)BaseDataObject[PdfName.D];
        return dashObject != null ? LineDash.Get(dashObject, null) : DefaultLineDash;
      }
      set
      {
        PdfArray dashObject = null;
        if(value != null)
        {
          dashObject = new PdfArray();
          foreach(double dashItem in value.DashArray)
          {dashObject.Add(PdfReal.Get(dashItem));}
        }
        BaseDataObject[PdfName.D] = dashObject;
      }
    }

    /**
      <summary>Gets/Sets the border style.</summary>
    */
    public StyleEnum Style
    {
      get
      {return ToStyleEnum((PdfName)BaseDataObject[PdfName.S]);}
      set
      {BaseDataObject[PdfName.S] = value != DefaultStyle ? ToCode(value) : null;}
    }

    /**
      <summary>Gets/Sets the border width in points.</summary>
    */
    public double Width
    {
      get
      {
        IPdfNumber widthObject = (IPdfNumber)BaseDataObject[PdfName.W];
        return widthObject != null ? widthObject.RawValue : DefaultWidth;
      }
      set
      {BaseDataObject[PdfName.W] = PdfReal.Get(value);}
    }
    #endregion
    #endregion
    #endregion
  }
}